DNA sequencing has revolutionized microbial ecology. There are two orthogonal approaches commonly used to explore the microbial universe: amplicon where a part of a single gene (usually the 16S rRNA gene) is amplified and sequenced and untargeted (“shotgun”) sequencing of all (“meta”) microbial genomes (“genomics”) where all the DNA is extracted and sequenced. 16S rRNA sequencing provides a rapid, portable and cheap method to identify bacterial organisms in a sample. Drawbacks of this technique are its low taxonomic resolution (typically only to the genus level), its inability to identify community members from other phylogenetic kingdoms or tell us what functional genes are present in a given community. Shotgun metagenomics allows for finer taxonomic resolution (species and/or strain level), cross kingdom identification and details about what genes microbial organisms are carrying in a given community.

<p align="center">
 <img src="./figures/shotgun.jpg" width="500">
</p>

Data analysis pipelines will always follow a pretty standard workflow. Once you understand this, then you will be able to apply these same ideas to other omics technologies using a different set of tools. 


                                           -> Assembly -> Taxonomic and functional annotation                                      
Raw data -> Quality check -> Preprocessing                                                      -> Data visualizaiton -> Analysis
                                           -> Taxonomic and functional annotation
